esi_leap/common/keystone.py

```diff
@@ -54,3 +57,38 @@ def get_project_uuid_from_ident(project_ident):
             # projects have unique names
             return projects[0].id
         raise exception.ProjectNoSuchName(name=project_ident)
+
+
+def refresh_project_list():
+    global _cached_project_list
+    _cached_project_list = get_keystone_client().projects.list()
+
+
+def get_project_list():
+    if _cached_project_list is None:
+        refresh_project_list()
+    return _cached_project_list
+
+
+@functools.lru_cache(maxsize=100)
+def search_project_list(search_value, search_key='id'):
+    project = next((p for p in get_project_list()
+                   if getattr(p, search_key) == search_value),
+                   None)
+    if project is None:
+        # refresh once to see if project is recently created
+        refresh_project_list()
+        project = next((p for p in get_project_list()
+                       if getattr(p, search_key) == search_value),
+                       None)
+    return project
+
+
+@functools.lru_cache(maxsize=100)
+def get_project_name(project_id):
+    project_name = ''
+    if project_id:
```

Comment:
Maybe Im not understanding fully how these functions are used, but why is there this first if project_id?

Reply:
Fair question! It's because not all of the leases/offers have associated project IDs for various things, and I figured that rather than put a bunch of 'if' statements within the API controller, I could just isolate it here.